core/app/models/spree/app_configuration.rb

```diff
@@ -136,6 +136,11 @@ class AppConfiguration < Preferences::Configuration
     #     charged (default: +14+)
     preference :expedited_exchanges_days_window, :integer, default: 14
 
+    # @!attribute [rw] generate_api_key_for_all_roles
+    #   @return [Boolean] Allow generating api key automatically for user
+    #   at role_user creation for all roles. (default: +false+)
+    preference :generate_api_key_for_all_roles, :boolean, default: false
```

Comment:
We could move the auto-generation to the user model and then get the benefit from this preference. That would give the solution you desire.